mlflow/store/s3_artifact_repo.py

```diff
@@ -35,7 +44,7 @@ def log_artifact(self, local_file, artifact_path=None):
         dest_path = posixpath.join(
             dest_path, os.path.basename(local_file))
         s3_client = self._get_s3_client()
-        s3_client.upload_file(local_file, bucket, dest_path)
+        s3_client.upload_file(local_file, bucket, dest_path, self.get_s3_file_upload_extra_args())
```

Comment:
I'm not too keen on having  json environment variables control the behavior because it's error prone and brittle.  We would need multiple env vars for each API call from S3 if we continued this pattern. 

Perhaps a better solution would be to extract out the upload into a separate interface and allow people to monkey patch the interface as necessary. Another would be to properly inject an interface to S3 for people to provide their own implementation.

Reply:
Right, although there's currently only three API calls (upload, list (paginate) and download), and upload seems the most demanding. See another PR on this here: https://github.com/mlflow/mlflow/pull/2461 - that one just aims at ACL.

Your idea for a solution - isn't that already possible like suggest here: https://github.com/mlflow/mlflow/issues/1662#issuecomment-592741183 ?

Reply:
Hello @andychow-db ,
I was facing similar issues, concretely need for ServerSideEncryption when uploading files:
[https://github.com/mlflow/mlflow/issues/1764](1764)

Sharing your sentiment regarding introduction of a ever-growing list of new environment variables, I also don't find monkey patching particularly pleasing. I've currently `gorilla` monkey patched SSE into the client interface, but needed to overwrite four different functions (on the repos and client, respectively) to pass arguments down and it feels pretty brittle and ugly.
I like your suggestion on having an interjectable interface for S3 (potentially with a default behavior that reflects the current state of affairs and documentation on how to customize) as that should be a lot more robust. If you would be interested in pursuing this way forward, I would be happy to support in whichever way I can.